text/0758-default-component-manager.md

```diff
@@ -0,0 +1,138 @@
+---
+Stage: Accepted
+Start Date: 2021-05-17
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s): Ember.js, Learning
+RFC PR: https://github.com/emberjs/rfcs/pull/758
+---
+
+# Default Component Manager
+
+## Summary
+
+Anything that can be in a template has its lifecycle managed by the manager pattern.
+Today, we have known managers for `@glimmer/component`, `@ember/helper`, etc.
+But what happens when the VM encounters an object for which there is no manager?,
+such as a plain function? This RFC explores and proposes a default behavior for
+those unknown scenarios when it comes to _components_.
+
+## Motivation
+
+The addon, [ember-could-get-used-to-this](https://github.com/pzuraq/ember-could-get-used-to-this)
+demonstrated that it's possible to use plain functions for helpers and modifiers.
+Components on the other hand, have had no such "plain" / "vanilla" support / adoption from addons.
+Maybe it's worth exploring what default behavior components should have.
+
+This could reduce the number of dependencies folks need to worry about when building components.
+Using native classes eliminates the reliance on a "framework thing" and further
+lessens the appearance of "magic through object-oriented inheritance".
```

Comment:
I *very strongly* disagree here. This is actually much *more* "magical": in this design, the framework calls a class constructor with specific arguments because it happens to be invoked in a specific way. It's *not* equivalent to function-based helpers or modifiers in my view:

- helpers are just functions which *you* pass all the arguments to
- modifiers are just functions which you pass all the arguments to as well, modulo the `element`—but notionally, sticking them on/applying them to the element *is* passing it that

To be sure, I can see a way to rationalize this in terms of saying that a component template is simply a specific kind of callback to the rendering engine (in a real sense that *is* what it is, of course), but it’s very strange notionally, and seems to me to move *away* from the actual value of explicitness that the Glimmer component subclass gives us.

I also don't think that there's anything particularly object-oriented about this use of inheritance, nor that it's in the realm of anti patterns that we *do* see from inheritance-based designs. There is effectively no *behavior* we're worried about here: just some boilerplate elimination via setting up the `owner` and `args`—and, importantly, doing it *safely* so that they don't get accidentally stomped later.

Insofar as one potential goal might be getting rid of the `willDestroy` hook, we can do that separately: it's perfectly reasonable to consider deprecating that from Glimmer `Component`, and it's also basically trivial to codemod to.